### PR TITLE
IR Changes

### PR DESCRIPTION
**Pull request by l-kent**
This is partially in response to discussions regarding #185 and #184 and does the following:

- merges MemoryStore into MemoryAssign, so that stores cannot be contained in other types of statements (which never made any sense to begin with)
- make BProcedure and BProcedure have default parameter values - this is just cleanup
- make Register take an Int as parameter instead of BitVecType - this is just nicer
- splits Memory into StackMemory and SharedMemory - this means we can match memory accesses on type to determine whether shared memory is being accessed, instead of matching the name "mem" vs. "stack". This will allow there to be multiple stack region variables in the future, for instance
- rename LocalAssign to Assign to avoid confusion about what 'local' means in this context
- a little bit of commenting to provide clarity on some of the more confusing parts of the IR

### CODE REVIEW

**Comment by l-kent** [comment]
Thinking about it a little further, Memory probably shouldn't be an Expr either

**Comment by l-kent** [comment]
These were generally relics of the IR hewing closely to BAP and/or Boogie for convenience

**Comment by l-kent** [comment]
The state register flags (ZF, NF, etc.) should probably be treated as globals and not locals because it probably isn't impossible the state of them in one procedure is relevant for the state in another procedure due to compiler optimisations?

**Comment by ailrst** [comment]
This is a good start, some further ideas:

- `MemoryAssign(var mem: Memory, var index: Expr, var value: Expr` assumes that the type of `index` and `value` is `BitVecType(size) with the size matching mem.addressSize and mem.valueSize respectively, this could be typechecked or we could make all of them more concrete. 
    - A more out-there idea is to have a pointer expression which carries the memory region with it, and have memory region be a parameter of the pointer/address rather than of the store/load operation, but this would mean describing pointer arithmetic.
- Now the only difference between `Assign` and `MemoryAssign` is whether the LHS is a scalar object or a map object which is clearer but it would help analyses if we could syntactically distinguish shared assigns from unshared assigns. For example the formalism has the three cases (unshared := unshared, shared := unshared, unshared := shared). With this modification we still need to do a full AST walk of the RHS expression to find variables and memory accesses and check whether they are shared or not. 
- There are still the implicit invariants below which would be nice to validate or have a comment
    1. That the RHS of `MemoryAssign` does not contain a load expression.
    2. That the RHS of `Assign` contains _at most one_ load expression.

**Comment by l-kent** [comment]
> MemoryAssign assumes that the type of index and value is BitVecType(size) with the size matching mem.addressSize and mem.valueSize respectively, this could be typechecked (as in with conditional throw) or we could make all of them more concrete.

I am not sure how you make these types more concrete. If these do not match, obviously invalid Boogie is generated which is why I have not yet bothered to implement type checking for these sizes. That could still be done but it's hardly a priority.

> Now the only difference between Assign and MemoryAssign is whether the LHS is a scalar object or a map object which is clearer but it would help analyses if we could syntactically distinguish shared assigns from unshared assigns. For example the formalism has the three cases (unshared := unshared, shared := unshared, unshared := shared). With this modification we still need to do a full AST walk of the RHS expression to find variables and memory accesses and check whether they are shared or not.

If a MemoryAssign is a write to shared state, then its mem parameter will be a SharedMemory.

Which analysis is all this important for?

> A more out-there idea is to have a pointer expression which carries the memory region with it, and have memory region be a parameter of the pointer/address rather than of the store/load operation, but this would mean describing pointer arithmetic.

This sounds needlessly convoluted for no real benefit.

**Comment by l-kent** [comment]
This is ready to merge @ailrst

**Review by ailrst — APPROVED** [review]
I think this change is good you can ignore my comments if you want.

File: src/main/scala/analysis/ANR.scala [hunk 1]
```diff
@@ -19,9 +19,9 @@ trait ANRAnalysis(cfg: ProgramCfg) {
 
   val first: Set[CfgNode] = Set(cfg.startNode)
 
-  private val stackPointer = Register("R31", BitVecType(64))
-  private val linkRegister = Register("R30", BitVecType(64))
-  private val framePointer = Register("R29", BitVecType(64))
+  private val stackPointer = Register("R31", 64)
+  private val linkRegister = Register("R30", 64)
+  private val framePointer = Register("R29", 64)
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/analysis/ANR.scala:24]: This PR doesn't make any changes to the scope. If you want to propose changes to how the scope works, you can do that separately. What you're suggesting would make the type system weaker though.

I did consider the implications for supporting an integer representation (as in SMACK) but that would require wider changes anyway, as everything else already assumes everything is bitvectors - the Memory case class already does for instance.

File: src/main/scala/analysis/ANR.scala [hunk 2]
```diff
@@ -35,12 +35,12 @@ trait ANRAnalysis(cfg: ProgramCfg) {
...
         m.diff(assert.body.variables)
       case memoryAssign: MemoryAssign =>
-        m.diff(memoryAssign.lhs.variables ++ memoryAssign.rhs.variables)
+        m.diff(memoryAssign.index.variables)
       case indirectCall: IndirectCall =>
         m - indirectCall.target
-      case localAssign: LocalAssign =>
-        m = m.diff(localAssign.rhs.variables)
-        if ignoreRegions.contains(localAssign.lhs) then m else m + localAssign.lhs
+      case assign: Assign =>
```
[the remark this answers is not in the document]
Comment (l-kent) [reply src/main/scala/analysis/ANR.scala:41]: As we've discussed before. this means creating a new temporary variable for nearly every load, adding a lot of bloat to the IR without actually improving anything.

File: src/main/scala/analysis/ANR.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (ailrst) [reply src/main/scala/analysis/ANR.scala:24]: Yeah I agree, if we want to lift registers to variables then we can lift registers to variables.

**Pull request merged**